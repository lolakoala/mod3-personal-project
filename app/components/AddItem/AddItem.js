import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AddBill from '../AddBill/AddBill.js';
import AddBulletin from '../AddBulletin/AddBulletin.js';
import AddChore from '../AddChore/AddChore.js';

class AddItem extends Component {
  constructor() {
    super();
    this.state = {
      item: ''
    };
  }

  updateItem = val => {
    this.setState({ item: val });
  }

  getTodaysDate = () => {
    let today = new Date();
    let dd = today.getDate();
    let mm = today.getMonth() + 1;
    let yyyy = today.getFullYear();
    if (dd < 10) {
      dd = '0' + dd;
    }
    if ( mm < 10) {
      mm='0'+mm;
    }
    return `${mm}/${dd}/${yyyy}`;
  }

  isActive = item => {
    return this.state.item === item ? 'active' : '';
  }

  renderButtons = () => {
    return (
      <div className='additem'>
        <button className={`add-bill-button ${this.isActive('bill')}`} onClick={() => this.updateItem('bill')}>Add Bill</button>
        <button className={`add-chore-button ${this.isActive('chore')}`} onClick={() => this.updateItem('chore')}>Add Chore</button>
        <button className={`add-bulletin-button ${this.isActive('bulletin')}`} onClick={() => this.updateItem('bulletin')}>Add Bulletin</button>
      </div>
    );
  }

  render() {
    const { usersHouse, currentUser, addBill, addBulletin, addChore } = this.props;
    return (
      <div className='additem-with-items'>
        {usersHouse.users ? this.renderButtons() : null}
        {this.state.item === 'bill' ?
          <AddBill usersHouse={usersHouse}
            currentUser={currentUser}
            addBill={addBill}
            getTodaysDate={this.getTodaysDate}
            updateItem={this.updateItem}/>
          : null}
        {this.state.item === 'bulletin' ?
          <AddBulletin usersHouse={usersHouse}
            currentUser={currentUser}
            addBulletin={addBulletin}
            getTodaysDate={this.getTodaysDate}
            updateItem={this.updateItem}/>
          : null}
        {this.state.item === 'chore' ?
          <AddChore usersHouse={usersHouse}
            currentUser={currentUser}
            addChore={addChore}
            getTodaysDate={this.getTodaysDate}
            updateItem={this.updateItem}/>
          : null}
      </div>
    );
  }
}

export default AddItem;

AddItem.propTypes = {
  usersHouse: PropTypes.object,
  currentUser: PropTypes.object,
  addBulletin: PropTypes.func,
  addBill: PropTypes.func,
  addChore: PropTypes.func
};
