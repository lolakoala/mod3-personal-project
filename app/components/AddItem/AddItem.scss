.additem {
  font-family: 'BenchNine', sans-serif;
  margin: 20px;
}

.additem button {
  font-family: 'BenchNine', sans-serif;
  border: none;
  background-color: #D83D59;
  width: 120px;
  height: 25px;
  color: #F4CCE2;
  font-size: 20px;
  border-radius: 25px;
  text-align: center;
  margin: 5px;
  display: block;
}

.additem button:hover,
.additem .active {
  cursor: pointer;
  border: 2px dashed #D83D59;
  background-color: #F4CCE2;
  color: #D83D59;
}

.add-item-with-items {
  padding-bottom: 150px;
}
