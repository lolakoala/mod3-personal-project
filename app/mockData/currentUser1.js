const currentUser1 = {
  id : "RVde1m9BhkQBLZlTJNsFlQZXqP43",
  name: "Amanda Brenner",
  house: "Dog9"
};

export default currentUser1;
