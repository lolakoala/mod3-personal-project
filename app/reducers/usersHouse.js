import firebase from '../firebase.js';

const usersHouse = (state = {}, action) => {
  const housesRef = firebase.database().ref('houses');
  switch (action.type) {
  case 'CREATE_HOUSE':
    housesRef.push(action.house);
    return action.house;
  case 'GET_HOUSE':
    const updatedHouse = Object.assign(action.usersHouse, { users: [...action.usersHouse.users, action.user] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updatedHouse);
    return updatedHouse;
  case 'LOGIN_SUCCESS':
    return action.usersHouse;
  case 'SIGNOUT':
    return {};
  case 'LEAVE_HOUSE':
    const deleteUser = action.usersHouse.users.filter(user => user.id !== action.currentUser.id);
    const houseWithoutUser = Object.assign(action.usersHouse, { users: deleteUser });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithoutUser);
    return {};
  case 'ADD_BILL':
    const takeOutFakeBills = action.usersHouse.bills.filter(bill => bill.title !== 'fake');
    const houseWithBill = Object.assign(action.usersHouse, { bills: [...takeOutFakeBills, action.bill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBill);
    return houseWithBill;
  case 'ADD_BULLETIN':
    const takeOutFakeBulletins = action.usersHouse.bulletins.filter(bulletin => bulletin.title !== 'fake');
    const houseWithBulletin = Object.assign(action.usersHouse, { bulletins: [...takeOutFakeBulletins, action.bulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithBulletin);
    return houseWithBulletin;
  case 'ADD_CHORE':
    const takeOutFakeChore = action.usersHouse.chores.filter(chore => chore.title !== 'fake');
    const houseWithChore = Object.assign(action.usersHouse, { chores: [...takeOutFakeChore, action.chore] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(houseWithChore);
    return houseWithChore;
  case 'ADD_READER':
    const oldBulletin = action.usersHouse.bulletins.find(bulletin => bulletin.id === action.bulletinId);
    if (oldBulletin.hasRead.includes(action.userId)) {
      return action.usersHouse;
    }
    const newBulletin = Object.assign(oldBulletin, { hasRead: [...oldBulletin.hasRead, action.userId] });
    const updateBulletins = action.usersHouse.bulletins.filter(bulletin => {
      return bulletin.id !== action.bulletinId;
    });
    const updateWithBulletin = Object.assign(action.usersHouse, { bulletins: [...updateBulletins, newBulletin] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBulletin);
    return updateWithBulletin;
  case 'MARK_PAID':
    const oldBill = action.usersHouse.bills.find(bill => bill.id === action.billId);
    const user = oldBill.allUsersTotals.find(user => user.id === action.userId);
    if (user.paid === true) {
      return action.usersHouse;
    }
    const newUser = Object.assign(user, { paid: true });
    const newAllUsersTotals = oldBill.allUsersTotals.filter(user => user.id !== action.userId);
    const newBill = Object.assign(oldBill, { allUsersTotals: [...newAllUsersTotals, newUser] });
    const updatedBills = action.usersHouse.bills.filter(bill => {
      return bill.id !== action.billId;
    });
    const updateWithBill = Object.assign(action.usersHouse, { bills: [...updatedBills, newBill] });
    firebase.database().ref("houses/" + action.usersHouse.houseKey).set(updateWithBill);
    return updateWithBill;
  default:
    return state;
  }
};

export default usersHouse;
