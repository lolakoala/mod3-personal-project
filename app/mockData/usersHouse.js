import currentUser1 from './currentUser1.js';
import currentUser2 from './currentUser2.js';

const bill1 = {
  allUsersTotals: [
    {
      id: currentUser1.id,
      paid: false,
      total: 50
    },
    {
      id: currentUser2.id,
      paid: false,
      total: 50
    }
  ],
  datePosted: "11/01/2017",
  details: "We used a lot of water.",
  duedate: "11/9/17",
  id: "b3019762705779",
  parsedDuedate: 1510210800000,
  postedBy: currentUser1,
  title: "Water Bill",
  total: "100"
};

const bulletin1 = {
  datePosted: "11/01/2017",
  details: "The dogs got out through a hole in the fence yesterday.",
  hasRead: [currentUser1.id],
  id: 1509567098747,
  postedBy: currentUser1,
  title: "Don't leave dogs out unattended."
};

const chore1 = {
  assignedTo: currentUser1.id,
  datePosted: "10/31/2017",
  details: "Empty the dishwasher.",
  done: false,
  id: "c1509485241076",
  postedBy: currentUser1,
  title: "Dishes",
  urgency: "low"
};

const usersHouse = {
  houseKey: "-KxoIBWy5UYRscFYdWty",
  bills: [bill1],
  bulletins: [bulletin1],
  chores: [chore1],
  houseCode: "woof9",
  houseName: "Dog9",
  users: [currentUser1, currentUser2]
};

export default usersHouse;
