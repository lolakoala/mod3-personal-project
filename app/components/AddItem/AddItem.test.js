import React from "react";
import AddItem from './AddItem.js';
import { shallow } from "enzyme";
import currentUser1 from '../../mockData/currentUser1.js';
import usersHouse from '../../mockData/usersHouse.js';

describe('AddItem component', () => {
  const mockFn = jest.fn();
  const wrapper = shallow(<AddItem usersHouse={usersHouse}
    currentUser={currentUser1}
    addChore={mockFn}
    addBill={mockFn}
    addBulletin={mockFn}/>);

  it('should exist', () => {
    expect(wrapper).toBeDefined();
  });

  it('should have a default state', () => {
    const defaultState = {
      item: ''
    };

    expect(wrapper.state()).toEqual(defaultState);
  });

  it('should have 3 buttons', () => {
    const buttons = wrapper.find('button');

    expect(buttons.length).toEqual(3);
  });

  it('should change state and show component when add bill button clicked', () => {
    const addBill = wrapper.find('.add-bill');

    addBill.simulate('click');
    expect(wrapper.state().item).toEqual('bill');
    expect(wrapper.find('AddBill').length).toEqual(1);
  });

  it('should change state and show component when add chore button clicked', () => {
    const addChore = wrapper.find('.add-chore');

    addChore.simulate('click');
    expect(wrapper.state().item).toEqual('chore');
    expect(wrapper.find('Chore').length).toEqual(1);
  });

  it('should change state and show component when add bulletin button clicked', () => {
    const addBulletin = wrapper.find('.add-bulletin');

    addBulletin.simulate('click');
    expect(wrapper.state().item).toEqual('bulletin');
    expect(wrapper.find('AddBulletin').length).toEqual(1);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
