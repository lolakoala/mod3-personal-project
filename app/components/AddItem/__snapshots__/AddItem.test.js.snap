// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItem component should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddItem
    addBill={[Function]}
    addBulletin={[Function]}
    addChore={[Function]}
    currentUser={
      Object {
        "house": "Dog9",
        "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
        "name": "Amanda Brenner",
      }
    }
    usersHouse={
      Object {
        "bills": Array [
          Object {
            "allUsersTotals": Array [
              Object {
                "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
                "paid": false,
                "total": 50,
              },
              Object {
                "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
                "paid": false,
                "total": 50,
              },
            ],
            "datePosted": "11/01/2017",
            "details": "We used a lot of water.",
            "duedate": "11/9/17",
            "id": "b3019762705779",
            "parsedDuedate": 1510210800000,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Water Bill",
            "total": "100",
          },
        ],
        "bulletins": Array [
          Object {
            "datePosted": "11/01/2017",
            "details": "The dogs got out through a hole in the fence yesterday.",
            "hasRead": Array [
              "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            ],
            "id": 1509567098747,
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Don't leave dogs out unattended.",
          },
        ],
        "chores": Array [
          Object {
            "assignedTo": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "datePosted": "10/31/2017",
            "details": "Empty the dishwasher.",
            "done": false,
            "id": "c1509485241076",
            "postedBy": Object {
              "house": "Dog9",
              "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
              "name": "Amanda Brenner",
            },
            "title": "Dishes",
            "urgency": "low",
          },
        ],
        "houseCode": "woof9",
        "houseKey": "-KxoIBWy5UYRscFYdWty",
        "houseName": "Dog9",
        "users": Array [
          Object {
            "house": "Dog9",
            "id": "RVde1m9BhkQBLZlTJNsFlQZXqP43",
            "name": "Amanda Brenner",
          },
          Object {
            "house": "Dog9",
            "id": "cFNdorl5ceRyMr9XGvPFRVEcw7I3",
            "name": "Lola Koala",
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="additem"
        >
          <button
            className="add-bill-button "
            onClick={[Function]}
          >
            Add Bill
          </button>
          <button
            className="add-chore-button "
            onClick={[Function]}
          >
            Add Chore
          </button>
          <button
            className="add-bulletin-button "
            onClick={[Function]}
          >
            Add Bulletin
          </button>
        </div>,
        null,
        null,
        null,
      ],
      "className": "additem-with-items",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="add-bill-button "
              onClick={[Function]}
            >
              Add Bill
            </button>,
            <button
              className="add-chore-button "
              onClick={[Function]}
            >
              Add Chore
            </button>,
            <button
              className="add-bulletin-button "
              onClick={[Function]}
            >
              Add Bulletin
            </button>,
          ],
          "className": "additem",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Add Bill",
              "className": "add-bill-button ",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add Bill",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Add Chore",
              "className": "add-chore-button ",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add Chore",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": null,
            "nodeType": "host",
            "props": Object {
              "children": "Add Bulletin",
              "className": "add-bulletin-button ",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Add Bulletin",
            "type": "button",
          },
        ],
        "type": "div",
      },
      null,
      null,
      null,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="additem"
          >
            <button
              className="add-bill-button "
              onClick={[Function]}
            >
              Add Bill
            </button>
            <button
              className="add-chore-button "
              onClick={[Function]}
            >
              Add Chore
            </button>
            <button
              className="add-bulletin-button "
              onClick={[Function]}
            >
              Add Bulletin
            </button>
          </div>,
          null,
          null,
          null,
        ],
        "className": "additem-with-items",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="add-bill-button "
                onClick={[Function]}
              >
                Add Bill
              </button>,
              <button
                className="add-chore-button "
                onClick={[Function]}
              >
                Add Chore
              </button>,
              <button
                className="add-bulletin-button "
                onClick={[Function]}
              >
                Add Bulletin
              </button>,
            ],
            "className": "additem",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Add Bill",
                "className": "add-bill-button ",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Bill",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Add Chore",
                "className": "add-chore-button ",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Chore",
              "type": "button",
            },
            Object {
              "instance": null,
              "key": null,
              "nodeType": "host",
              "props": Object {
                "children": "Add Bulletin",
                "className": "add-bulletin-button ",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Add Bulletin",
              "type": "button",
            },
          ],
          "type": "div",
        },
        null,
        null,
        null,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
